### PR TITLE
syncer: record and display synchronization position gaps

### PR DESCRIPTION
**Pull request by holys**
#111

### CODE REVIEW

File: syncer/db.go [hunk 1]
```diff
@@ -531,6 +531,31 @@ func isRetryableError(err error) bool {
 	return true
 }
 
+func isAccessDeniedError(err error) bool {
+	return isXXXError(err, tmysql.ErrSpecificAccessDenied)
+}
+
+func isXXXError(err error, code uint16) bool {
```
Comment (GregoryIan) [inline syncer/db.go]: :laughing: what is XXX？

Comment (holys) [inline syncer/db.go]: Need better naming

File: syncer/db.go [hunk 2]
```diff
@@ -673,3 +698,56 @@ func closeDBs(dbs ...*sql.DB) {
 		}
 	}
 }
+
+func showMasterStatus(db *sql.DB) (gmysql.Position, error) {
```
Comment (GregoryIan) [inline syncer/db.go]: return gtidSet infomation although it's null

File: syncer/db.go [hunk 3]
```diff
@@ -673,3 +698,56 @@ func closeDBs(dbs ...*sql.DB) {
...
+	rows, err := querySQL(db, "SHOW MASTER STATUS")
+	if err != nil {
+		return gmysql.Position{}, errors.Trace(err)
```
Comment (GregoryIan) [inline syncer/db.go]: `return nil, errors.Trace(err)`

File: syncer/db.go [hunk 4]
```diff
@@ -673,3 +698,56 @@ func closeDBs(dbs ...*sql.DB) {
...
+	}
+	defer rows.Close()
+
+	rowColumns, err := rows.Columns()
+	if err != nil {
+		return gmysql.Position{}, errors.Trace(err)
```
Comment (GregoryIan) [inline syncer/db.go]: ditto

File: syncer/db.go [hunk 5]
```diff
@@ -673,3 +698,56 @@ func closeDBs(dbs ...*sql.DB) {
...
+		+-------------------+----------+--------------+------------------+-------------------+
+		| master-bin.000001 |      548 |              |                  |                   |
+		+-------------------+----------+--------------+------------------+-------------------+
+	*/
+
+	var file string
+	var position string
+
+	for rows.Next() {
+		data := make([]sql.RawBytes, len(rowColumns))
```
Comment (GregoryIan) [inline syncer/db.go]: why not use ```var (
  file string
  pos xxx
)```

File: syncer/db.go [hunk 6]
```diff
@@ -673,3 +698,56 @@ func closeDBs(dbs ...*sql.DB) {
...
+			return gmysql.Position{}, errors.Trace(err)
+		}
+
+		file = string(data[0])
+		position = string(data[1])
+	}
+
+	posUint64, err := strconv.ParseUint(position, 10, 64)
+	if err != nil {
+		return gmysql.Position{}, errors.Trace(err)
```
Comment (GregoryIan) [inline syncer/db.go]: ditto

File: syncer/metrics.go [hunk 7]
```diff
@@ -52,20 +61,20 @@ var (
 			Help:      "total number of sql retryies",
 		}, []string{"type"})
 
-	binlogMetaPos = prometheus.NewGauge(
+	binlogMeta = prometheus.NewGaugeVec(
```
Comment (GregoryIan) [inline syncer/metrics.go]: split binlog-file and binlog-pos is better

File: syncer/db.go (hunk 3 above)
Comment (holys) [inline syncer/db.go]: No, you can't return nil as type for that it's not a pointer

Comment (GregoryIan) [inline syncer/db.go]: so, why don't return a pointer

Comment (holys) [inline syncer/db.go]: For code convention consistency. go-mysql/replication use Position type, and syncer/db.go use Position type.

File: syncer/metrics.go (hunk 7 above)
Comment (holys) [inline syncer/metrics.go]: Show your reasons, please.

Comment (GregoryIan) [inline syncer/metrics.go]: how would you display them in prometheus or granfana,  is it comparable between pos and file in one graph? spit it or add one more type label to distinguish them

File: syncer/db.go (hunk 3 above)
Comment (GregoryIan) [inline syncer/db.go]: nothing about convention consistency，it's personal taste.
`gmysql.Position` is small, so u can do,  same to `TableName` and `RouteRule`.
 u should make them consistent, then as u like

File: syncer/metrics.go (hunk 7 above)
Comment (holys) [inline syncer/metrics.go]: Whatever, if you want to compare pos or file, use two separated graphs, but one struct is enough.

image

image

Comment (GregoryIan) [inline syncer/metrics.go]: 🤣 The  [time series](https://en.wikipedia.org/wiki/Time_series) values are preferably of the same type. it's comparable

Comment (andelf) [inline syncer/metrics.go]: A metric name should represent the same logical thing-being-measured across all label dimensions.

ref: https://prometheus.io/docs/practices/naming/

**Comment by holys** [comment]
@GregoryIan  any problem?

**Comment by GregoryIan** [comment]
let #113 merge firstly

File: syncer/db.go [hunk 8]
```diff
@@ -543,6 +543,31 @@ func isRetryableError(err error) bool {
...
+	var e error
+	for {
+		e = errors.Cause(err)
+		if err == e {
+			break
+		}
+		err = e
+	}
+	mysqlErr, ok := err.(*mysql.MySQLError)
+	if ok {
```
Comment (GregoryIan) [inline syncer/db.go]: a suggest `return ok && mysqlErr.Number == code`

File: syncer/db_test.go [hunk 9]
```diff
@@ -24,6 +27,25 @@ import (
...
+func (s *testDBSuite) SetUpSuite(c *C) {
+	host := "127.0.0.1"
+	port := 3306
+	user := "root"
+	password := ""
+
+	var err error
+	dbAddr := fmt.Sprintf("%s:%s@tcp(%s:%d)/?charset=utf8", user, password, host, port)
+	s.db, err = sql.Open("mysql", dbAddr)
+	if err != nil {
```
Comment (GregoryIan) [inline syncer/db_test.go]: `c.Assert(err, IsNil)`

Comment (GregoryIan) [inline syncer/db_test.go]: u can use `testSyncerSuite`, don't need `testDBSuite`
`func (s *testSyncerSuite) TestIsRetryableError`
`func (s * testSyncerSuite) TestShowMasterStatus`

File: syncer/metrics.go (hunk 7 above)
Comment (GregoryIan) [inline syncer/metrics.go]: address it like @andelf  said

File: syncer/meta.go [hunk 10]
```diff
@@ -152,3 +152,13 @@ func (lm *LocalMeta) String() string {
 	gtids := lm.GTID()
 	return fmt.Sprintf("binlog-name = %s, binlog-pos = %d, binlog-gtid = %v", pos.Name, pos.Pos, gtids)
 }
+
+// MasterMeta represents binlog meta information of master
+type MasterMeta struct {
```
Comment (GregoryIan) [inline syncer/meta.go]: I don't think u need define MasterMeta, simpify code

File: syncer/db_test.go (hunk 9 above)
Comment (holys) [inline syncer/db_test.go]: testSyncerSuite do  `reset master`, and then  `getMasterStatus`  failed.

Comment (GregoryIan) [inline syncer/db_test.go]: modify `testSyncerSuite `, I means u don't need two test suits

File: syncer/meta.go (hunk 10 above)
Comment (holys) [inline syncer/meta.go]: I think it's OK.

Comment (GregoryIan) [inline syncer/meta.go]: syncer only uses in `printStatus`(except `reSyncGtid`),  use  position and gtid directly is also simple. If `MasterMeta` contains many field, I would think it's better, but no now.  as `syncerBinlogPos  ...` that defined
 in `syncer/metrics.go` it's also useless, why not use it directly?

**Comment by GregoryIan** [comment]
Rest LGTM

File: syncer/meta.go (hunk 10 above)
Comment (holys) [inline syncer/meta.go]: `MasterMeta ` is easy for print( it implements Stringer)

**Comment by iamxy** [comment]
LGTM

**Pull request merged**